torch/_inductor/codecache.py

```diff
@@ -147,79 +147,170 @@ def is_gcc():
     return re.search(r"(gcc|g\+\+)", cpp_compiler())
 
 
-class _SupportedVecIsa(enum.Enum):
-    AVX512 = 1
-    AVX2 = 2
-    INVALID = -1
+class VecISA(object):
+    _bit_width: int
+    _macro: str
+    _arch_flags: str
+    _dtype_nelements: Dict[torch.dtype, int]
+
+    # TorchInductor CPU vectorization reuses PyTorch vectorization utility functions
+    # Hence, TorchInductor would depend on Sleef* to accelerate mathematical functions
+    # like exp, pow, sin, cos and etc.
+    # But PyTorch and TorchInductor might use different compilers to build code. If
+    # PyTorch uses gcc-7/g++-7 to build the release package, the libtorch_cpu.so
+    # will not expose the Sleef* AVX512 symbols since gcc-7/g++-7 cannot pass
+    # avx512 check in CMake - FindAVX.cmake. But TorchInductor install the latest
+    # gcc/g++ compiler by default while it could support the AVX512 compilation.
+    # Therefore, there would be a conflict sleef version between PyTorch and
+    # TorchInductor. Hence, we dry-run the following code to check whether current
+    # HW platform and PyTorch both could support AVX512 or AVX2. And suppose ARM
+    # also needs the logic
+    _avx_code = """
+#if defined(CPU_CAPABILITY_AVX512) || defined(CPU_CAPABILITY_AVX2)
+#include <ATen/cpu/vec/functional.h>
+#include <ATen/cpu/vec/vec.h>
+#endif
+
+__attribute__((aligned(64))) float in_out_ptr0[16] = {0.0};
+
+int main() {
+    auto tmp0 = at::vec::Vectorized<float>(1);
+    auto tmp1 = tmp0.exp();
+    tmp1.store(in_out_ptr0);
+    return 0;
+}
+"""
+
+    def bit_width(self):
+        return self._bit_width
+
+    def nelements(self, dtype: torch.dtype = torch.float):
+        return self._dtype_nelements[dtype]
+
+    def build_macro(self):
+        return self._macro
+
+    def build_arch_flags(self):
+        return self._arch_flags
+
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+    @functools.lru_cache(None)
+    def __bool__(self):
+        key, input_path = write(VecISA._avx_code, "cpp", extra="")
+        from filelock import FileLock
+
+        lock_dir = get_lock_dir()
+        lock = FileLock(os.path.join(lock_dir, key + ".lock"), timeout=LOCK_TIMEOUT)
```

Comment:
Could we reuse more of the standard compile path for this?  

It might also make sense to cache the result to disk similar to the triton autotuner.  I worry this will hurt compile times.

Reply:
Hi Janson, I encountered a performance issue If I use the standard compiler path to load the generated library to check ISA. The performance of MT5ForConditionalGeneration dropped from 4.01x to 1.77x for multi-thread. I haven't investigated the root cause as I'm focusing on optimizing the outlier models. Hence, this PR creates another process to dry-run a python script to load the generated library to isolate the execution environment.

I will file an issue to track your comments and polish it later. Are you okay with that?

Reply:
OK. In that case, `lru_cache(1)` is more appropriate?

Reply:
It always returns the same result. I'm also curious about it because it only impacts `MT5ForConditionalGeneration`.

Reply:
> OK. In that case, lru_cache(1) is more appropriate?

Hi @jgong5 , `lru_cache(1)` will evict the previous `VecISA` instance. Since there are three global `VecISA` instances, we should not use `lru_cache(1)` from the performance perspective.

Reply:
Ah, yes, you are right. Sorry about the confusion.